File: bluetooth/src/main/java/com/ankerboxmanager/bluetooth/scaner/BluetoothScanner.java
//package com.ankerboxmanager.bluetooth.scaner;
//
//import java.util.HashSet;
//import java.util.Set;
//
//import com.ankerboxmanager.bluetooth.IBluetooth;
//import com.ankerboxmanager.bluetooth.bean.AnkerBoxBluetoothDevice;
//
//import android.bluetooth.BluetoothDevice;
//import android.content.Context;
//import android.util.Log;
//
///**
// * Created on 14/05/2018.
// *
// * @author lx
// */
//public abstract class BluetoothScanner {
//
//    private static final String TAG = IBluetooth.TAG;
//
//    public BluetoothScanner(Context context) {
//        this.context = context;
//        mDeviceFoundListeners = new HashSet<>();
//        mFoundedBluetoothDevices = new HashSet<>();
//    }
//
//    protected Context context;
//
//    public abstract void startScan(String[] serviceUUIDs, IBluetooth.AsyncReturnCallback asyncReturnCallback);
//
//    public abstract void stopScan(IBluetooth.AsyncReturnCallback asyncReturnCallback);
//
//    private Set<IBluetooth.DeviceFoundListener> mDeviceFoundListeners;
//
//    private Set<AnkerBoxBluetoothDevice> mFoundedBluetoothDevices;
//
//    public boolean addDeviceFoundListener(IBluetooth.DeviceFoundListener listener) {
//        return mDeviceFoundListeners.add(listener);
//    }
//
//    public boolean removeDeviceFoundListener(IBluetooth.DeviceFoundListener listener) {
//        return mDeviceFoundListeners.remove(listener);
//    }
//
//    public void destroy() {
//        mDeviceFoundListeners.clear();
//        mFoundedBluetoothDevices.clear();
//        context = null;
//    }
//
//    public Set<AnkerBoxBluetoothDevice> getFoundedBluetoothDevices() {
//        return mFoundedBluetoothDevices;
//    }
//
//    void onNewDeviceFound(BluetoothDevice bondedDevice) {
//        AnkerBoxBluetoothDevice abDevice = buildDevice(bondedDevice);
//        mFoundedBluetoothDevices.add(abDevice);
//        for (IBluetooth.DeviceFoundListener listener : mDeviceFoundListeners) {
//            listener.onNewDeviceFound(abDevice);
//        }
//        Log.d(TAG, this.getClass().getSimpleName() + ", onNewDeviceFound, abDevice=" + abDevice);
//    }
//
//    public AnkerBoxBluetoothDevice buildDevice(BluetoothDevice bondedDevice) {
//        AnkerBoxBluetoothDevice device = new AnkerBoxBluetoothDevice();
//        device.name = bondedDevice.getName();
//        device.deviceId = bondedDevice.getAddress();
//        return device;
//    }
//
//}
//

File: bluetooth/src/main/kotlin/com/ankerboxmanger/bluetooth/scaner/BluetoothScanner.kt
package com.ankerboxmanger.bluetooth.scaner

import java.util.HashSet

import android.bluetooth.BluetoothDevice
import android.content.Context
import android.util.Log
import com.ankerboxmanger.bluetooth.IBluetooth
import com.ankerboxmanger.bluetooth.bean.AnkerBoxBluetoothDevice

import com.ankerboxmanger.bluetooth.IBluetooth.Companion.TAG

/**
 * Created on 14/05/2018.
 *
 * @author lx
 */
abstract class BluetoothScanner(protected var context: Context) {

    private val deviceFoundListeners: MutableSet<IBluetooth.DeviceFoundListener>
            by lazy(LazyThreadSafetyMode.NONE) { HashSet<IBluetooth.DeviceFoundListener>() }

    val foundedBluetoothDevices: MutableSet<AnkerBoxBluetoothDevice> by lazy { HashSet<AnkerBoxBluetoothDevice>() }

    abstract fun startScan(serviceUUIDs: Array<String>?, asyncReturnCallback: IBluetooth.AsyncReturnCallback?)

    abstract fun stopScan(asyncReturnCallback: IBluetooth.AsyncReturnCallback?)

    fun addDeviceFoundListener(listener: IBluetooth.DeviceFoundListener): Boolean {
        return deviceFoundListeners.add(listener)
    }

    fun removeDeviceFoundListener(listener: IBluetooth.DeviceFoundListener): Boolean {
        return deviceFoundListeners.remove(listener)
    }

    open fun destroy() {
        deviceFoundListeners.clear()
        foundedBluetoothDevices.clear()
    }

    internal fun onNewDeviceFound(bondedDevice: BluetoothDevice) {
        val abDevice = buildDevice(bondedDevice)
        foundedBluetoothDevices.add(abDevice)
        for (listener in deviceFoundListeners) {
            listener.onNewDeviceFound(abDevice)
        }
        Log.d(TAG, this.javaClass.simpleName + ", onNewDeviceFound, abDevice=" + abDevice)
    }

    fun buildDevice(bondedDevice: BluetoothDevice): AnkerBoxBluetoothDevice {
        val device = AnkerBoxBluetoothDevice(bondedDevice.name, bondedDevice.address)
        return device
    }

}

